File: android/app/src/main/java/com/itcorner/moodleandroid/data/mappers/GradeMapper.kt
package com.itcorner.moodleandroid.data.mappers

import com.itcorner.moodleandroid.data.dto.GradeItemDto
import com.itcorner.moodleandroid.domain.GradeItem

object GradeMapper {
    fun map(dto: GradeItemDto): GradeItem {
        val normalizedGrade = dto.gradeformatted
            ?.trim()
            ?.takeIf { it.isNotEmpty() && it != "-" }
            ?: "Not graded"

        val normalizedPercentage = dto.percentageformatted
            ?.trim()
            ?.takeIf { it.isNotEmpty() && it != "-" }

        return GradeItem(
            id = dto.id,
            itemName = dto.itemname?.takeIf { it.isNotBlank() } ?: "Unnamed item",
            gradeText = normalizedGrade,
            percentageText = normalizedPercentage
        )
    }
}

File: ios/MoodleAssessmentApp/MoodleAssessmentApp/Data/Mapping/GradeMapper.swift
import Foundation

enum GradeMapper {
    static func map(_ dto: GradeItemDTO) -> GradeItem {
        let name = dto.itemname ?? "Untitled Item"
        let gradeText = normalizedGradeText(dto.gradeformatted) ?? formattedRawGrade(from: dto) ?? "Not graded"

        return GradeItem(
            id: dto.id,
            itemName: name,
            gradeText: gradeText,
            percentageText: dto.percentageformatted
        )
    }

    private static func formattedRawGrade(from dto: GradeItemDTO) -> String? {
        guard let raw = dto.graderaw else { return nil }

        if dto.grademin != nil, let max = dto.grademax {
            return "\(raw.cleanNumber) / \(max.cleanNumber)"
        }

        return raw.cleanNumber
    }

    private static func normalizedGradeText(_ value: String?) -> String? {
        guard let raw = value?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
            return nil
        }

        if raw == "-" {
            return nil
        }

        return raw
    }
}

private extension Double {
    var cleanNumber: String {
        truncatingRemainder(dividingBy: 1) == 0 ? String(format: "%.0f", self) : String(format: "%.2f", self)
    }
}
